generator/place_processor.cpp

```diff
@@ -45,18 +46,49 @@ double GetRadiusM(ftypes::LocalityType const & type)
 template <typename T>
 bool IsWorsePlace(T const & left, T const & right)
 {
-  auto const rankL = left.GetRank();
-  auto const rankR = right.GetRank();
-  auto const levelL = ftype::GetLevel(GetPlaceType(left));
-  auto const levelR = ftype::GetLevel(GetPlaceType(right));
-  auto const langCntL = left.GetMultilangName().CountLangs();
-  auto const langCntR = right.GetMultilangName().CountLangs();
-  auto const isPointL = left.IsPoint();
-  auto const isPointR = right.IsPoint();
-  auto const boxAreaL = left.GetLimitRect().Area();
-  auto const boxAreaR = right.GetLimitRect().Area();
-  return std::tie(rankL, levelL, langCntL, isPointL, boxAreaL) <
-      std::tie(rankR, levelR, langCntR, isPointR, boxAreaR);
+  double constexpr kRankCoeff = 1.0;
+  double constexpr kIsCapitalCoeff = 1.0;
+  double constexpr kLangsCountCoeff = 1.0;
+  double constexpr kIsPointCoeff = 0.1;
+  double constexpr kIsNodeCoeff = 0.1;
+  double constexpr kAreaCoeff = 0.1;
+
+  auto const normaliseRank = [](uint8_t rank) {
+    return static_cast<double>(rank) / static_cast<double>(std::numeric_limits<uint8_t>::max());
+  };
+  auto const normaliseLangsCount = [](uint8_t langsCount) {
+    return static_cast<double>(langsCount) /
+           static_cast<double>(StringUtf8Multilang::kMaxSupportedLanguages);
+  };
+  auto const normaliseArea = [](double area) {
+    double const kMaxAreaM2 = 2e9;
+    area = base::Clamp(area, 0.0, kMaxAreaM2);
+    return area / kMaxAreaM2;
+  };
+
+  static_assert(kRankCoeff >= 0, "");
+  static_assert(kIsCapitalCoeff >= 0, "");
+  static_assert(kLangsCountCoeff >= 0, "");
+
+  auto const getScore = [&](auto const place) {
+    auto const rank = place.GetRank();
+    auto const isCapital = ftypes::IsCapitalChecker::Instance()(GetPlaceType(place));
+    auto const langsCount = place.GetMultilangName().CountLangs();
+    auto const isPoint = place.IsPoint();
+
+    auto const type = place.GetMostGenericOsmId().GetType();
+    auto const isNode = (type == base::GeoObjectId::Type::OsmNode) || (type == base::GeoObjectId::Type::ObsoleteOsmNode);
+
+    auto const limitRect = place.GetLimitRect();
+    auto const area = mercator::DistanceOnEarth(limitRect.LeftBottom(), limitRect.LeftTop()) *
+                      mercator::DistanceOnEarth(limitRect.LeftBottom(), limitRect.RightBottom());
+
+    return kRankCoeff * normaliseRank(rank) + kIsCapitalCoeff * (isCapital ? 1.0 : 0.0) +
+           kLangsCountCoeff * normaliseLangsCount(langsCount) +
+           kIsPointCoeff * (isPoint ? 1.0 : 0.0) + kIsNodeCoeff * (isNode ? 1.0 : 0.0) + kAreaCoeff * normaliseArea(area);
```

Comment:
nit: I'd override clang-format here and left one summand per line

Reply:
done